### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by philkunz** [node-fetch/node-fetch#1618]
Our cluster shows a memory leak when upgrading to 3.x.x
2.x.x shows no memory leak.
We are not using any keep alive logic.

Note: The issue becomes apparent when handling 1000s of requests / minute pretty quickly. The services simply die with out of memory errors. Note that the memory rises when node memory is monitored for the whole pid group of the node.js process, but not when just taking a node.js memory heap snapshot.

**Comment by stalkerg** [comment]
I don't know about 2.x.x, but on 3.3.0 I have huge memory leak

**Comment by steckhelena** [comment]
We are experiencing the same thing on a service using version 3.3.0 in production and doing `httpts` requests. Sometimes we have to do a burst of requests and we can clearly see the heap usage increasing fast during that time and not falling down again after(so the GC is not collecting some objects created by `node-fetch`). We can also observe it when doing requests slowly, albeit at a slower pace, thus why it might not be affecting many people. Inspecting locally we can see that a `TLSSocket` is responsible for a lot of memory usage and they don't seem to get cleaned up.

**Comment by philkunz** [comment]
just ditch this node-fetch and use @adobe/fetch instead which works smoothly without any problems.

**Comment by jimmywarting** [comment]
Do you have any code example?
are you consuming the response also or only making a request?

**Comment by steckhelena** [comment]
I am making the request and checking for the response status. Strangely enough in a simple while loop it works without leaks, so I couldn't find a simple reproducible example. Also on version 2.6.7, or some other libraries, it does not leak, which is why it makes me think there is some problem in the 3.x.x version.

I just looked at my grafana too, and can see a lot of open file descriptors and the bumps in the open file descriptors happens at the same time as the bump in heap memory usage.

**Comment by stalkerg** [comment]
My case was fixed here https://github.com/nodejs/undici/pull/1824 maybe node-fetch has something similar.

**Comment by jimmywarting** [comment]
@philkunz it's just no fun when ppl just go ahead and recommend other competitive libraries and more or less talk down about node-fetch, never do that in any open source project.
`@adobe/fetch` has nothing to do with this particular memory issue, it's unrelated. And their codebase is different the ours. it dose not help us figuring out what the memory issue really is.

**Comment by jimmywarting** [comment]
> I am making the request and checking for the response status. Strangely enough in a simple while loop it works without leaks

Do you consume the body or not? do you use signals?

undici uses weakref's which we can't really use just yet b/c of the NodeJS version we have to support. we would like to use it but we can't unless we make a breaking release.

**Comment by philkunz** [comment]
@jimmywarting: Open Source means the users are owed nothing and the authors are owed nothing just because it is Open Source. There is a license that governs attribution. And you can delete my comments all you want. For making this a better world and for getting shit done -> @adobe/fetch solves a problem. Btw. I also maintain Open Source packages under MIT license. If there is better package for someone, I don't care, or I'll use it myself, or I'll copy the code if the license allows it with proper attribution.

**Comment by steckhelena** [comment]
> Do you consume the body or not? do you use signals?

No, I don't consume the body. I use an AbortController, get an AbortSignal from there, and pass it into the fetch call. Example of how I am calling fetch:

```js
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'some-other-headers': header,
        },
        body: payload,
        signal: controller.signal,
        redirect: 'manual',
        follow: 0,
      })
```

And then the only thing I do with this response is check for the code and throw an error if it is `code >= 300` and `code < 200`. I have `setTimeout` for aborting the request after a while and clear them in case of errors and also on success.

**Comment by EpicWink** [comment]
> No, I don't consume the body

Likely duplicates #83
